src/native/libs/System.Globalization.Native/pal_icushim_static.c

```diff
@@ -98,57 +116,356 @@ static int32_t load_icu_data(void* pData)
     }
 }
 
-int32_t GlobalizationNative_LoadICUData(const char* path)
+#if defined(USE_APPLE_DECOMPRESSION)
+static int
+apple_cache_directory(char path[static (PATH_MAX + 1)])
+{
+    char buf[PATH_MAX + 1];
+    sysdir_search_path_enumeration_state st = sysdir_start_search_path_enumeration(SYSDIR_DIRECTORY_CACHES, SYSDIR_DOMAIN_MASK_USER);
+    st = sysdir_get_next_search_path_enumeration(st, buf);
+    if (!st)
+    {
+        log_shim_error("apple_cache_directory: sysdir_get_next_search_path_enumeration did not yield a path");
+        return -1;
+    }
+    size_t path_len = strlen(buf);
+    size_t ret_len = path_len + 2 /* '/' + '\0' */;
+    const char *home = "";
+    size_t home_len = 0;
+    if (buf[0] == '~' && buf[1] == '/') {
+        home = getenv("HOME");
+        if (home == NULL)
+        {
+            log_shim_error("apple_cache_directory: cache directory begins with ~ but $HOME is not set");
+            return -1;
+        }
+        home_len = strlen(home);
+        ret_len += home_len + 1;
+    }
+    snprintf(path, PATH_MAX, "%s%s", home, &buf[1]);
+    return 0;
+}
+
+#define APPLE_DECOMPRESSION_BUF_SIZE 16384
+
+static int
+apple_decompress_to_fd(int dst_fd, size_t *dst_len, const char *src_buf, size_t src_len)
+{
+    int cs_init = 0;
+    compression_stream cs = { 0 };
+    uint8_t buf[APPLE_DECOMPRESSION_BUF_SIZE];
+    size_t dst_size = 0;
+
+    compression_status status = compression_stream_init(&cs, COMPRESSION_STREAM_DECODE, COMPRESSION_LZFSE);
+    if (status == COMPRESSION_STATUS_ERROR)
+    {
+        log_shim_error("apple_decompress_to_fd: Failed to initialize decompression stream.");
+        goto error;
+    }
+    cs_init = 1;
+    cs.src_ptr = (const uint8_t *) src_buf;
+    cs.src_size = src_len;
+    cs.dst_ptr = buf;
+    cs.dst_size = APPLE_DECOMPRESSION_BUF_SIZE;
+
+    int flags = 0;
+    while (status == COMPRESSION_STATUS_OK)
+    {
+        status = compression_stream_process(&cs, flags);
+        if (status == COMPRESSION_STATUS_ERROR)
+        {
+            log_shim_error("apple_decompress_to_fd: Error while decompressing.");
+            goto error;
+        }
+
+        size_t bytes_to_write = APPLE_DECOMPRESSION_BUF_SIZE - cs.dst_size;
+        uint8_t *write_cursor = buf;
+        dst_size += bytes_to_write;
+        while (bytes_to_write > 0) {
+            ssize_t bytes_written = write(dst_fd, write_cursor, bytes_to_write);
+            int last_error = errno;
+            if (bytes_written == -1)
+            {
+                if (last_error == EINTR) continue;
+                log_shim_error("apple_decompress_to_fd: Error during write().");
```

Comment:
What will the app do if the write fails (or anything else here fails)? Will it exit/crash, or will it fall back to executing, but without any ICU available (whatever that means)?

One reasonable cause for failure is the device running out of disk space.

Reply:
If ICU fails to load, the app will terminate with a call to `Environment::FailFast`

Reply:
> What will the app do if the write fails (or anything else here fails)? Will it exit/crash, or will it fall back to executing, but without any ICU available (whatever that means)?
> 
> One reasonable cause for failure is the device running out of disk space.

From testing: the app will continue without ICU data available, so internationalization-related functions will return no data or will fail, exactly as if icudt.dat is not available. We could try decompressing into private memory if writing fails, and only fall back to this degraded failure mode if both disk and memory are unavailable.